[ProxyTensor] dedupe symbolic shapes in tracing

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115441
* __->__ #116158

Dedupes symbolic shapes in proxy tensor tracing. Reusing the existing sym shape avoids inserting spurious sym_size calls, which can interfere with pattern matching and graph passes.

eellison committed:
[ProxyTensor] dedupe symbolic shapes in tracing

[ghstack-poisoned]

eellison committed:
Update on "[ProxyTensor] dedupe symbolic shapes in tracing"


Dedupes symbolic shapes in proxy tensor tracing. Reusing the existing sym shape avoids inserting spurious sym_size calls, which can interfere with pattern matching and graph passes. 



[ghstack-poisoned]

ezyang commented:
I think as a reviewer I would definitely appreciated a more detailed PR description.

ezyang commented on test/test_proxy_tensor.py:
```diff
@@ -1019,6 +1019,25 @@ def forward(self, x_1, y_1):
+            self.assertExpectedInline(r, """\
+def forward(self, s0_1, s1_1, x_1, y_1):
+    empty = torch.ops.aten.empty.memory_format([s0_1], device = device(type='cpu'), pin_memory = False)
+    return ((s0_1, s1_1), empty)""")
```
ok cool, I buy it

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -933,6 +1026,12 @@ def make_fx(f,
+        decomposition_table = {
+            **decomposition_table,
+            torch.ops.aten.sym_numel.default: torch._decomp.decompositions.sym_numel
+        }
```
What's going on here?

eellison replied:
I want the sym_numels to be expressed in terms of inputs. This is reduces uses on tensors, but is also needed because I had a scenario where a SymNumel was first used in a tensor in the backward that ends up int he backward and then later used in a tensor that is in the forward. 

Is there a better way of decomposing sym_numel? I tried adding some logic [https://github.com/pytorch/pytorch/pull/116158/files#diff-6960ce90e7162c0953b1ca07e92e7f0f2f6ba63b427b42df593e20cc6a096bb7R254](here) but it added detach nodes into the graph for reasons I couldnt understand.

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -534,7 +627,7 @@ def wrapped(*proxies):
         out = pytree.tree_map_only(
             (SymInt, SymFloat, SymBool),
-            lambda t: get_proxy_slot(t.node, tracer)(),
+            lambda t: get_proxy_slot(t, tracer)(),
```
looks like some sort of refactor

eellison replied:
This is because we're no longer using the node to hash, but the SymInt itself

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -74,6 +74,70 @@ def decompose(decomposition_table):
+        return self.sym_obj.node.expr == value.sym_obj.node.expr
+
+
+class _SymHashingDict:
```
This implementation seems way stronger than what your test is exercising. What you are doing here is something like if I compute s0 + s1 multiple times, I will discover and CSE all of these into a single expression. The test, OTOH, seems to imply that if you did the CSE only on expressions that were merely Symbol, that would have been sufficient. Do you actually care about more complex CSE? Because if you don't, I think this code gets a lot simpler.

eellison replied:
I think the more complex CSE here is good.. I'm not sure how much simpler it gets if I don't attempt to it. I could not rely on the Sympy hash implementation but otherwise should be pretty similar ?

eellison replied:
I can add more tests

ezyang reviewed (approved):
This seems functionally correct so I am approving but I think it is more complicated than it needs to be.

eellison committed:
Update on "[ProxyTensor] dedupe symbolic shapes in tracing"


Dedupes symbolic shapes in proxy tensor tracing. Reusing the existing sym shape avoids inserting spurious sym_size calls, which can interfere with pattern matching and graph passes. 



[ghstack-poisoned]

eellison committed:
Update on "[ProxyTensor] dedupe symbolic shapes in tracing"


Dedupes symbolic shapes in proxy tensor tracing. Reusing the existing sym shape avoids inserting spurious sym_size calls, which can interfere with pattern matching and graph passes. 



[ghstack-poisoned]

eellison commented:
@pytorchbot merge
